internal/blobstore/s3store.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package blobstore // import "gopkg.in/juju/charmstore.v5-unstable/internal/blobstore"
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"log"
+
+	"gopkg.in/amz.v3/aws"
+	"gopkg.in/amz.v3/s3"
+	"gopkg.in/errgo.v1"
+)
+
+type s3Store struct {
+	bucket string
+	getS3  func() *s3.S3
+}
+
+// NewS3 createa a new S3 backed blobstore Store
+func NewS3(pc *ProviderConfig) *Store {
+	return &Store{newS3(pc)}
+}
+
+func newS3(pc *ProviderConfig) *s3Store {
+	getter := getS3(pc)
+	s := &s3Store{
+		bucket: pc.BucketName,
+		getS3:  getter,
+	}
+	s.createBucket()
+	return s
+}
+
+func (s *s3Store) createBucket() {
+	bucket, err := s.getS3().Bucket(s.bucket)
+	if err != nil { //this only happens on invalid bucket name that is when a name has any of /:@
```

Comment:
I always worry about statements like this when it isn't explicitly in a functions documentation.